File: backjoon_review/exhaustiveSearch/9019.py
import sys
from collections import deque

input = sys.stdin.readline

def bfs(start):
    q = deque()
    q.append((start, ""))
    isVisited[start] = True

    while q:
        num, command = q.popleft()

        if num == endNum:
            return command

        print("num : %d" %num)
        nextNum = num * 2
        if nextNum > 9999:
            nextNum = nextNum % 10000

        print(nextNum)
        if isVisited[nextNum] == False:
            q.append((nextNum, command + "D"))
            isVisited[nextNum] = True



        #S
        nextNum = num -1
        if nextNum == -1:
            nextNum = 9999

        print(nextNum)
        if isVisited[nextNum] == False:
            q.append((nextNum, command + "S"))
            isVisited[nextNum] = True


        #L
        nextNum = int((num % 1000) * 10 + num / 1000)
        print(nextNum)
        if isVisited[nextNum] == False:
            q.append((nextNum, command + "L"))
            isVisited[nextNum] = True

        #R
        nextNum = int((num % 10) * 1000 + num / 10)
        if isVisited[nextNum] == False:
            q.append((nextNum, command + "R"))
            isVisited[nextNum] = True

        print(nextNum)

for t in range(int(input())):
    startNum , endNum = map(int,input().split())
    isVisited = [False] * 10001
    print(bfs(startNum))

File: baekjoon_js/exhaustiveSearch/9019.js
("use strict");

const fs = require("fs");

const stdin = (process.platform === "linux"
  ? fs.readFileSync("/dev/stdin").toString()
  : `3
1234 3412
1000 1
1 16
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

let T = Number(input());
let num, target;
let isVisited;
while (T--) {
  [num, target] = input().split(" ").map(Number);
  isVisited = Array(10001).fill(false);
  console.log(BFS(num));
}

// let a = [];
// a = a + "hi";
// console.log(a);

function BFS(start) {
  q = [];
  q.push([start, []]);

  while (q.length > 0) {
    let [nowNum, history] = q.shift();
    isVisited[nowNum] = true;
    // console.log(nowNum, history);
    if (nowNum === target) return history;

    let commandD = nowNum * 2;
    if (commandD > 9999) commandD = commandD % 10000;
    if (isVisited[commandD] === false) {
      q.push([commandD, history + "D"]);
      //   isVisited[commandD] = true;
    }

    let commandS = nowNum - 1;
    if (commandS === 0) commandS = 9999;
    if (isVisited[commandS] === false) {
      q.push([commandS, history + "S"]);
      //   isVisited[commandS] = true;
    }

    // let commandL = "";
    // commandL += String(nowNum)[1];
    // commandL += String(nowNum)[2];
    // commandL += String(nowNum)[3];
    // commandL += String(nowNum)[0];

    let commandL = Number(
      String((nowNum % 1000) * 10 + parseInt(nowNum / 1000))
    );
    // console.log(nowNum, commandL);
    if (isVisited[commandL] === false) {
      q.push([commandL, history + "L"]);
      //   isVisited[Number(commandL)] = true;
    }
    // let commandR = "";
    // commandR += String(nowNum)[3];
    // commandR += String(nowNum)[0];
    // commandR += String(nowNum)[1];
    // commandR += String(nowNum)[2];

    let commandR = Number(String((nowNum % 10) * 1000 + parseInt(nowNum / 10)));
    if (isVisited[commandR] === false) {
      q.push([commandR, history + "R"]);
      //   isVisited[Number(commandR)] = true;
    }
  }
}
